part-4-intro.md rewrite - oliver (#1055)
rust-lang/rustc-dev-guide: src/part-4-intro.md

# Part 4: Analysis

This part discusses analyses which the compiler uses to check various properties
of source code and to inform later compilation stages. Typically the analyses

> prestented here is what people mean when they refer to "Rust's type system".

Reviewer: ```suggestion
presented here is what people mean when they refer to "Rust's type system".
```

Also, the "the analyses ... is what people mean" sounds a bit funny, but so does using "are". Is there a better phrase this?

> This includes inference and checking of types, the trait and type systems
> representation, and the borrow checker generally. These analyses are separated

Reviewer: ```suggestion
The analyses include inference and type-checking, trait solving, and borrow-checking.
This section also discusses rustc's representation of types. These analyses are separated
```

throughout various parts of the compilation process and use meaningfully
different Intermediate Representations (IR). For example, type checking happens
on the HIR, while borrow checking happens on the MIR.

This section of the guide discusses all of these analyses which are the heart of

> Rust's novel approach to ownership which balances type system *expressivity with
> usability*, and which represents the integration of decades of programming

Reviewer: ```suggestion
Rust's novel approach to ownership, which balances type system *expressivity* with
*usability*, and which represents the integration of decades of programming
```
